### PR TITLE
[Filesystem] Workaround cannot dumpFile into "protected" folders on Windows

### LINKED ISSUE #39496: [Filesystem] Cannot dumpFile into "protected" folders on Windows
**Symfony version(s) affected**: 5.2.0

**Description**

Usage of `Symfony\Component\Filesystem\Filesystem::dumpFile` failes on Windows when trying to dump e.g. into the personal downloads directory under `C:\Users\*\Downloads` due to failing of the writeability check for the target folder even though an actual write succeeds.

**How to reproduce**

```php
<?php

require_once __DIR__ . '/vendor/autoload.php';

$targetFilePath = 'C:/Users/*user*/Downloads/file.ext';

var_dump(is_writable(dirname($targetFilePath))); // bool(false)
var_dump(file_put_contents($targetFilePath, 'foo') !== false); // bool(true)

$filesystem = new Symfony\Component\Filesystem\Filesystem();
$filesystem->dumpFile($targetFilePath, 'bar'); // exception
```

**Possible Solution**

Is more of an upstream-problem with PHP's implementation of `is_writable` (see e.g. https://bugs.php.net/bug.php?id=68926) but it might make sense to just skip the check on Windows to provide a fix within the symfony component.

The root cause is probably the Microsoft Universal Windows Platform app security model.

**Additional context**  

Testing on Windows 10 1904 with
```
PHP 7.4.12 (cli) (built: Oct 27 2020 17:18:47) ( ZTS Visual C++ 2017 x64 )
Copyright (c) The PHP Group
Zend Engine v3.4.0, Copyright (c) Zend Technologies
```

### PR DESCRIPTION
**Pull request by arnegroskurth**
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #39496
| License       | MIT
| Doc PR        |

### CODE REVIEW

File: src/Symfony/Component/Filesystem/Tests/FilesystemTest.php [hunk 1]
```diff
@@ -1751,6 +1751,26 @@ public function testCopyShouldKeepExecutionPermission()
         $this->assertFilePermissions(767, $targetFilePath);
     }
 
+    public function testDumpToProtectedDirectory()
+    {
+        if (PHP_OS_FAMILY !== 'Windows') {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Filesystem/Tests/FilesystemTest.php]: please see how we check for windows in this file

File: src/Symfony/Component/Filesystem/Tests/FilesystemTest.php [hunk 2]
```diff
@@ -1751,6 +1751,26 @@ public function testCopyShouldKeepExecutionPermission()
...
+            $this->markTestSkipped('This test is specific to Windows.');
+        }
+
+        if (($userProfilePath = getenv('USERPROFILE')) === false || !is_dir($userProfilePath)) {
+            throw new \RuntimeException('Failed to retrieve user profile path.');
+        }
+
+        // todo: is this path valid for non-english installations?
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Filesystem/Tests/FilesystemTest.php]: should be removed, we don't need to answer this until someone reports there is an issue here

File: src/Symfony/Component/Filesystem/Tests/FilesystemTest.php [hunk 3]
```diff
@@ -1751,6 +1751,26 @@ public function testCopyShouldKeepExecutionPermission()
...
+        $targetPath = implode(DIRECTORY_SEPARATOR, [$userProfilePath, 'Downloads', '__test_file.ext']);
+
+        try {
+            $this->filesystem->dumpFile($targetPath, 'foobar');
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Filesystem/Tests/FilesystemTest.php:1768]: this part is missing an assertion

**Review by nicolas-grekas — COMMENTED** [review]
I'm fine with the change, the error message will change, that's the only side effect

**Review by nicolas-grekas — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @arnegroskurth.

**Pull request merged**